File: Windows/32bit/Matlab/histomode.m

% Demo for access to HydraHarp 400 Hardware via HHLIB.DLL v 3.0.
% The program performs a measurement based on hard coded settings.
% The resulting histogram (65536 channels) is stored in an ASCII output file.
%
% Michael Wahl, PicoQuant, September 2014


% Constants from hhdefin.h

REQLIBVER   =     '3.0';    % this is the version this program expects
MAXDEVNUM   =         8;
MAXHISTBINS =     65536;	 % number of histogram channels
MAXLENCODE  =         6;	 % max histogram length is 65536	
MAXBINSTEPS	=    	 26;
MODE_HIST   =         0;
MODE_T2	    =         2;
MODE_T3	    =         3;

FLAG_OVERFLOW = hex2dec('0001');

ZCMIN		  =           0;		% mV
ZCMAX		  =          20;		% mV
DISCRMIN	  =           0;	   % mV
DISCRMAX	  =         800;	   % mV
OFFSETMIN	  =           0;		 % ps
OFFSETMAX	  =  1000000000;	    % ps

ACQTMIN		  =           1;		 % ms
ACQTMAX		  =   360000000;	    % ms  (100*60*60*1000ms = 100h)

% Errorcodes from errorcodes.h

HH_ERROR_DEVICE_OPEN_FAIL		 = -1;

% Settings for the measurement, Adapt to your setup!

SyncCFDZeroX  = 10;      %  you can change this
SyncCFDLevel  = 50;      %  you can change this
InputCFDZeroX = 10;      %  you can change this
InputCFDLevel = 50;      %  you can change this
SyncDiv       = 8;       %  you can change this
Binning       = 0;       %  you can change this
Tacq          = 1000;    %  you can change this      
    
fprintf('\nHydraHarp 400 HHLib Demo Application             PicoQuant 2014\n');

if (~libisloaded('HHlib'))    
    %Attention: The header file name given below is case sensitive and must
    %be spelled exactly the same as the actual name on disk except the file 
    %extension. 
    %Wrong case will apparently do the load successfully but you will not
    %be able to access the library!
    %The alias is used to provide a fixed spelling for any further access via
    %calllib() etc, which is also case sensitive.
    loadlibrary('hhlib.dll',   'hhlib.h', 'alias', 'HHlib'); % Windows 32 bit
    % loadlibrary('hhlib64.dll', 'hhlib.h', 'alias', 'HHlib'); % Windows 64 bit
    % loadlibrary('/usr/local/lib/hh400/hhlib.so', 'hhlib.h', 'alias', 'HHlib'); % Linux 32 bit
    % loadlibrary('/usr/local/lib64/hh400/hhlib.so', 'hhlib.h', 'alias', 'HHlib'); % Linux 64 bit
else
    fprintf('Note: HHlib was already loaded\n');
end;

if (libisloaded('HHlib'))
    fprintf('HHlib opened successfully\n');
    %libfunctionsview('HHlib'); %use this to test for proper loading
else
    fprintf('Could not open HHlib\n');
    return;
end;
    
LibVersion    = blanks(8); %enough length!
LibVersionPtr = libpointer('cstring', LibVersion);

[ret, LibVersion] = calllib('HHlib', 'HH_GetLibraryVersion', LibVersionPtr);
if (ret<0)
    fprintf('Error in GetLibVersion. Aborted.\n');
    err = HH_GETLIBVERSION_ERROR;
else
	fprintf('HHLib version is %s\n', LibVersion);
end;

if ~strcmp(LibVersion,REQLIBVER)
    fprintf('This program requires HHLib version %s\n', REQLIBVER);
    return;
end;

fid = fopen('histomode.out','w');
if (fid<0)
    fprintf('Cannot open output file\n');
    return;
end;

 fprintf(fid,'Binning           : %ld\n',Binning);
 fprintf(fid,'AcquisitionTime   : %ld\n',Tacq);
 fprintf(fid,'SyncDivider       : %ld\n',SyncDiv);
 fprintf(fid,'SyncCFDZeroCross  : %ld\n',SyncCFDZeroX);
 fprintf(fid,'SyncCFDLevel      : %ld\n',SyncCFDLevel);
 fprintf(fid,'InputCFDZeroCross : %ld\n',InputCFDZeroX);
 fprintf(fid,'InputCFDLevel1    : %ld\n',InputCFDLevel);


fprintf('\nSearching for HydraHarp devices...');

dev = [];
found = 0;
Serial     = blanks(8); %enough length!
SerialPtr  = libpointer('cstring', Serial);
ErrorStr   = blanks(40); %enough length!
ErrorPtr   = libpointer('cstring', ErrorStr);

for i=0:MAXDEVNUM-1
    [ret, Serial] = calllib('HHlib', 'HH_OpenDevice', i, SerialPtr);
    if (ret==0)       % Grab any HydraHarp we successfully opened
        fprintf('\n  %1d        S/N %s', i, Serial);
        found = found+1;            
        dev(found)=i; %keep index to devices we may want to use
    else
        if(ret==HH_ERROR_DEVICE_OPEN_FAIL)
            fprintf('\n  %1d        no device', i);
        else 
            [ret, ErrorStr] = calllib('HHlib', 'HH_GetErrorString', ErrorPtr, ret);
            fprintf('\n  %1d        %s', i,ErrorStr);
        end;
	end;
end;
    
% In this demo we will use the first HydraHarp device we found, i.e. dev(1).
% If you have nultiple HydraHarp devices you could also check for a specific 
% serial number, so that you always know which physical device you are talking to.

if (found<1)
	fprintf('\nNo device available. Aborted.\n');
	return; 
end;

fprintf('\nUsing device #%1d',dev(1));
fprintf('\nInitializing the device...');

[ret] = calllib('HHlib', 'HH_Initialize', dev(1), MODE_HIST, 0); 
if(ret<0)
	fprintf('\nHH_Initialize error %d. Aborted.\n',ret);
    closedev;
	return;
end; 

%this is only for information
Model      = blanks(16); %enough length!
Partno     = blanks(8); %enough length!
Version    = blanks(8); %enough length!
ModelPtr   = libpointer('cstring', Model);
PartnoPtr  = libpointer('cstring', Partno);
VersionPtr = libpointer('cstring', Version);

[ret, Model, Partno] = calllib('HHlib', 'HH_GetHardwareInfo', dev(1), ModelPtr, PartnoPtr, VersionPtr);
if (ret<0)
    fprintf('\nHH_GetHardwareInfo error %1d. Aborted.\n',ret);
    closedev;
	return;
else
	fprintf('\nFound model %s part number %s version %s', Model, Partno, Version);             
end;

NumInpChannels = int32(0);
NumChPtr = libpointer('int32Ptr', NumInpChannels);
[ret,NumInpChannels] = calllib('HHlib', 'HH_GetNumOfInputChannels', dev(1), NumChPtr); 
if (ret<0)
    fprintf('\nHH_GetNumOfInputChannels error %1d. Aborted.\n',ret);
    closedev;
	return;
else
	fprintf('\nDevice has %i input channels.', NumInpChannels);             
end;

fprintf('\nCalibrating ...');
[ret] = calllib('HHlib', 'HH_Calibrate', dev(1));
if (ret<0)
    fprintf('\nHH_Calibrate error %1d. Aborted.\n',ret);
    closedev;
    return;
end;
   
[ret] = calllib('HHlib', 'HH_SetSyncDiv', dev(1), SyncDiv);
if (ret<0)
    fprintf('\nHH_SetSyncDiv error %1d. Aborted.\n',ret);
    closedev;
    return;
end;

[ret] = calllib('HHlib', 'HH_SetSyncCFD', dev(1), SyncCFDLevel, SyncCFDZeroX);
if (ret<0)
    fprintf('\nHH_SyncSetCFD error %ld. Aborted.\n', ret);
    closedev;
    return;
end;

 
[ret] = calllib('HHlib', 'HH_SetSyncChannelOffset', dev(1), 0);
if (ret<0)
   fprintf('\nHH_SetSyncChannelOffset error %ld. Aborted.\n', ret);
   closedev;
   return;
end; 

for i=0:NumInpChannels-1 % we use the same input settings for all channels
   
    [ret] = calllib('HHlib', 'HH_SetInputCFD', dev(1), i, InputCFDLevel, InputCFDZeroX);
        if (ret<0)
        fprintf('\nHH_SetInputCFD error %ld. Aborted.\n', ret);
        closedev;
        return;
        end;   
   
    [ret] = calllib('HHlib', 'HH_SetInputChannelOffset', dev(1), i, 0);
        if (ret<0)
        fprintf('\nHH_SetInputChannelOffset error %ld. Aborted.\n', ret);
        closedev;
        return;
        end;
   
end;

HistLen = int32(0);
HistLenPtr = libpointer('int32Ptr', HistLen);
[ret, HistLen] = calllib('HHlib', 'HH_SetHistoLen', dev(1), MAXLENCODE, HistLenPtr);
if (ret<0)
    fprintf('\nHH_SetHistoLen error %ld. Aborted.\n', ret);
    closedev;
    return;
end;

[ret] = calllib('HHlib', 'HH_SetBinning', dev(1), Binning);
if (ret<0)
    fprintf('\nHH_SetBinning error %ld. Aborted.\n', ret);
    closedev;
    return;
end;
   
[ret] = calllib('HHlib', 'HH_SetOffset', dev(1), 0);
if (ret<0)
    fprintf('\nHH_SetOffset error %ld. Aborted.\n', ret);
    closedev;
    return;
end;

ret = calllib('HHlib', 'HH_SetStopOverflow', dev(1), 0, 10000); %for example only 
if (ret<0)
    fprintf('\nHH_SetStopOverflow error %ld. Aborted.\n', ret);
    closedev;
    return;
 end;
 
Resolution = 0;
ResolutionPtr = libpointer('doublePtr', Resolution);
[ret, Resolution] = calllib('HHlib', 'HH_GetResolution', dev(1), ResolutionPtr);
if (ret<0)
    fprintf('\nHH_GetResolution error %ld. Aborted.\n', ret);
    closedev;
    return;
 end;
 fprintf('\nResolution=%1dps', Resolution);


pause(0.4); % after Init or SetSyncDiv you must allow 400 ms for valid new count rates
            % otherwise you get new values every 100 ms

% From here you can repeat the measurement (with the same settings)


Syncrate = 0;
SyncratePtr = libpointer('int32Ptr', Syncrate);
[ret, Syncrate] = calllib('HHlib', 'HH_GetSyncRate', dev(1), SyncratePtr);
if (ret<0)
    fprintf('\nHH_GetSyncRate error %ld. Aborted.\n', ret);
    closedev;
    return;
end;
fprintf('\nSyncrate=%1d/s', Syncrate);
 
for i=0:NumInpChannels-1
    
	Countrate = 0;
	CountratePtr = libpointer('int32Ptr', Countrate);
	[ret, Countrate] = calllib('HHlib', 'HH_GetCountRate', dev(1), i, CountratePtr);
	if (ret<0)
   	fprintf('\nHH_GetCountRate error %ld. Aborted.\n', ret);
   	closedev;
   	return;
	end;
	fprintf('\nCountrate%1d=%1d/s ', i, Countrate);
   
end;

%new from v1.2: after getting the count rates you can check for warnings
Warnings = 0;
WarningsPtr = libpointer('int32Ptr', Warnings);
[ret, Warnings] = calllib('HHlib', 'HH_GetWarnings', dev(1), WarningsPtr);
if (ret<0)
    fprintf('\nHH_GetWarnings error %ld. Aborted.\n',ret);
    closedev;
    return;
end;
if (Warnings~=0)
    Warningstext = blanks(16384); %enough length!
    WtextPtr     = libpointer('cstring', Warningstext);
    [ret, Warningstext] = calllib('HHlib', 'HH_GetWarningsText', dev(1), WtextPtr, Warnings);
    fprintf('\n\n%s',Warningstext);
end;


ret = calllib('HHlib', 'HH_ClearHistMem', dev(1));    
if (ret<0)
    fprintf('\nHH_ClearHistMem error %ld. Aborted.\n', ret);
    closedev;
    return;
end;
        
ret = calllib('HHlib', 'HH_StartMeas', dev(1),Tacq); 
if (ret<0)
    fprintf('\nHH_StartMeas error %ld. Aborted.\n', ret);
    closedev;
    return;
end;
         
fprintf('\nMeasuring for %1d milliseconds...',Tacq);
        
ctcdone = int32(0);
ctcdonePtr = libpointer('int32Ptr', ctcdone);
while (ctcdone==0)
    [ret,ctcdone] = calllib('HHlib', 'HH_CTCStatus', dev(1), ctcdonePtr);
end;    
         
ret = calllib('HHlib', 'HH_StopMeas', dev(1)); 
if (ret<0)
    fprintf('\nHH_StopMeas error %ld. Aborted.\n', ret);
    closedev;
    return;
end;
        
countsbuffer  = uint32(zeros(NumInpChannels,MAXHISTBINS));
for i=0:NumInpChannels-1  
    bufferptr = libpointer('uint32Ptr', countsbuffer(i+1,:));
    [ret,countsbuffer(i+1,:)] = calllib('HHlib', 'HH_GetHistogram', dev(1), bufferptr, i, 0); 
    if (ret<0)
        fprintf('\nHH_GetHistogram error %ld. Aborted.\n', ret);
        closedev;
        return;
    end;
end;

flags = int32(0);
flagsPtr = libpointer('int32Ptr', flags);
[ret,flags] = calllib('HHlib', 'HH_GetFlags', dev(1), flagsPtr);
if (ret<0)
    fprintf('\nHH_GetFlags error %ld. Aborted.\n', ret);
    closedev;
    return;
end;
        
if(bitand(uint32(flags),FLAG_OVERFLOW)) 
    fprintf('  Overflow.');
end;    

Integralcount = sum(countsbuffer');        
fprintf('\nTotalCount=%1d', Integralcount);

fprintf('\nSaving file...');

for (i=1:MAXHISTBINS)
    fprintf(fid,'%7d ', countsbuffer(:,i));
    fprintf(fid,'\n');
end;    

plot(countsbuffer')

fprintf('\nData is in histomode.out ');

closedev;
    
if(fid>0) 
    fclose(fid);
end;

File: Windows/32bit/C/histomode/histomode.c
/************************************************************************

  Demo access to HydraHarp 400 Hardware via HHLib v.3.0.
  The program performs a measurement based on hardcoded settings.
  The resulting histogram (65536 channels) is stored in an ASCII output file.

  Michael Wahl, PicoQuant GmbH, April 2022

  Note: This is a console application (i.e. run in Windows cmd box)

  Note: At the API level channel numbers are indexed 0..N-1 
		where N is the number of channels the device has.

  Tested with the following compilers:

  - MinGW 2.0.0 (Windows 32 bit)
  - MinGW-W64 4.3.5 (Windows 64 bit)
  - MS Visual C++ 6.0 (Windows 32 bit)
  - MS Visual C++ 2019 (Windows 32 and 64 bit)
  - gcc 7.5.0 and 9.3.0 (Linux 64 bit)

************************************************************************/

#ifdef _WIN32
#include <windows.h>
#include <dos.h>
#include <conio.h>
#else
#include <unistd.h>
#define Sleep(msec) usleep(msec *1000)
#endif

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "hhdefin.h"
#include "hhlib.h"
#include "errorcodes.h"

unsigned int counts[HHMAXINPCHAN][MAXHISTLEN];


int main(int argc, char *argv[])
{
  int dev[MAXDEVNUM];
  int found = 0;
  FILE * fpout;
  int retcode;
  int ctcstatus;
  char LIB_Version[8];
  char HW_Model[16];
  char HW_Partno[8];
  char HW_Version[8];
  char HW_Serial[8];
  char Errorstring[40];
  int NumChannels;
  int HistLen;
  int Binning = 0;	//you can change this
  int Offset = 0;
  int Tacq = 1000;	//Measurement time in millisec, you can change this
  int SyncDivider = 1;	//you can change this 
  int SyncCFDZeroCross = 10;	//you can change this (in mV)
  int SyncCFDLevel = 50;	//you can change this (in mV)
  int SyncChannelOffset = -5000;	//you can change this (in ps, like a cable delay)
  int InputCFDZeroCross = 10;	//you can change this (in mV)
  int InputCFDLevel = 50;	//you can change this (in mV)
  int InputChannelOffset = 0;	//you can change this (in ps, like a cable delay)
  double Resolution;
  int Syncrate;
  int Countrate;
  double Integralcount;
  int i, j;
  int flags;
  int warnings;
  char warningstext[16384];	//must have 16384 bytest text buffer
  char cmd = 0;


  printf("\nHydraHarp 400 HHLib Demo Application        M. Wahl, PicoQuant GmbH, 2022");
  printf("\n~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~");
  HH_GetLibraryVersion(LIB_Version);
  printf("\nLibrary version is %s", LIB_Version);
  if (strncmp(LIB_Version, LIB_VERSION, sizeof(LIB_VERSION)) != 0)
    printf("\nWarning: The application was built for version %s.", LIB_VERSION);

  if ((fpout = fopen("histomode.out", "w")) == NULL)
  {
    printf("\ncannot open output file\n");
    goto ex;
  }

  fprintf(fpout, "Binning           : %d\n", Binning);
  fprintf(fpout, "Offset            : %d\n", Offset);
  fprintf(fpout, "AcquisitionTime   : %d\n", Tacq);
  fprintf(fpout, "SyncDivider       : %d\n", SyncDivider);
  fprintf(fpout, "SyncCFDZeroCross  : %d\n", SyncCFDZeroCross);
  fprintf(fpout, "SyncCFDLevel      : %d\n", SyncCFDLevel);
  fprintf(fpout, "InputCFDZeroCross : %d\n", InputCFDZeroCross);
  fprintf(fpout, "InputCFDLevel     : %d\n", InputCFDLevel);

  printf("\nSearching for HydraHarp devices...");
  printf("\nDevidx     Status");

  for (i = 0; i < MAXDEVNUM; i++)
  {
    retcode = HH_OpenDevice(i, HW_Serial);
    if (retcode == 0)	//Grab any HydraHarp we can open
    {
      printf("\n  %1d        S/N %s", i, HW_Serial);
      dev[found] = i;	//keep index to devices we want to use
      found++;
    }
    else
    {
      if (retcode == HH_ERROR_DEVICE_OPEN_FAIL)
        printf("\n  %1d        no device", i);
      else
      {
        HH_GetErrorString(Errorstring, retcode);
        printf("\n  %1d        %s", i, Errorstring);
      }
    }
  }

  //In this demo we will use the first HydraHarp device we find, i.e. dev[0].
  //You can also use multiple devices in parallel.
  //You can also check for specific serial numbers, so that you always know 
  //which physical device you are talking to.

  if (found < 1)
  {
    printf("\nNo device available.");
    goto ex;
  }

  printf("\nUsing device #%1d", dev[0]);
  printf("\nInitializing the device...");

  fflush(stdout);

  retcode = HH_Initialize(dev[0], MODE_HIST, 0);  //Histo mode with internal clock
  if (retcode < 0)
  {
    printf("\nHH_Initialize error %d. Aborted.\n", retcode);
    goto ex;
  }

  retcode = HH_GetHardwareInfo(dev[0], HW_Model, HW_Partno, HW_Version);  //only for information
  if (retcode < 0)
  {
    printf("\nHH_GetHardwareInfo error %d. Aborted.\n", retcode);
    goto ex;
  }
  else
    printf("\nFound Model %s Part no %s Version %s", HW_Model, HW_Partno, HW_Version);

  retcode = HH_GetNumOfInputChannels(dev[0], &NumChannels);
  if (retcode < 0)
  {
    printf("\nHH_GetNumOfInputChannels error %d. Aborted.\n", retcode);
    goto ex;
  }
  else
    printf("\nDevice has %i input channels.", NumChannels);

  fflush(stdout);

  printf("\nCalibrating...");
  retcode = HH_Calibrate(dev[0]);
  if (retcode < 0)
  {
    printf("\nCalibration Error %d. Aborted.\n", retcode);
    goto ex;
  }

  fflush(stdout);

  retcode = HH_SetSyncDiv(dev[0], SyncDivider);
  if (retcode < 0)
  {
    printf("\nPH_SetSyncDiv error %d. Aborted.\n", retcode);
    goto ex;
  }

  retcode = HH_SetSyncCFD(dev[0], SyncCFDLevel, SyncCFDZeroCross);
  if (retcode < 0)
  {
    printf("\nHH_SetSyncCFD error %d. Aborted.\n", retcode);
    goto ex;
  }

  retcode = HH_SetSyncChannelOffset(dev[0], SyncChannelOffset);
  if (retcode < 0)
  {
    printf("\nHH_SetSyncChannelOffset error %d. Aborted.\n", retcode);
    goto ex;
  }

  for (i = 0; i < NumChannels; i++)	// we use the same input settings for all channels, you can change this
  {
    retcode = HH_SetInputCFD(dev[0], i, InputCFDLevel, InputCFDZeroCross);
    if (retcode < 0)
    {
      printf("\nHH_SetInputCFD error %d. Aborted.\n", retcode);
      goto ex;
    }

    retcode = HH_SetInputChannelOffset(dev[0], i, InputChannelOffset);
    if (retcode < 0)
    {
      printf("\nHH_SetInputChannelOffset error %d. Aborted.\n", retcode);
      goto ex;
    }
  }

  retcode = HH_SetHistoLen(dev[0], MAXLENCODE, &HistLen);
  if (retcode < 0)
  {
    printf("\nHH_SetHistoLen error %d. Aborted.\n", retcode);
    goto ex;
  }
  printf("\nHistogram length is %d", HistLen);

  retcode = HH_SetBinning(dev[0], Binning);
  if (retcode < 0)
  {
    printf("\nHH_SetBinning error %d. Aborted.\n", retcode);
    goto ex;
  }

  retcode = HH_SetOffset(dev[0], Offset);
  if (retcode < 0)
  {
    printf("\nHH_SetOffset error %d. Aborted.\n", retcode);
    goto ex;
  }

  retcode = HH_GetResolution(dev[0], &Resolution);
  if (retcode < 0)
  {
    printf("\nHH_GetResolution error %d. Aborted.\n", retcode);
    goto ex;
  }

  printf("\nResolution is %1.1lfps\n", Resolution);

  fflush(stdout);

  //Note: after Init or SetSyncDiv you must allow >400 ms for valid  count rate readings
  //Otherwise you get new values after every 100 ms
  Sleep(400);

  retcode = HH_GetSyncRate(dev[0], &Syncrate);
  if (retcode < 0)
  {
    printf("\nHH_GetSyncRate error %d. Aborted.\n", retcode);
    goto ex;
  }
  printf("\nSyncrate=%1d/s", Syncrate);

  for (i = 0; i < NumChannels; i++)	// for all channels
  {
    retcode = HH_GetCountRate(dev[0], i, &Countrate);
    if (retcode < 0)
    {
      printf("\nHH_GetCountRate error %d. Aborted.\n", retcode);
      goto ex;
    }
    printf("\nCountrate[%1d]=%1d/s", i, Countrate);
  }

  printf("\n");

  //after getting the count rates you can check for warnings
  retcode = HH_GetWarnings(dev[0], &warnings);
  if (retcode < 0)
  {
    printf("\nHH_GetWarnings error %d. Aborted.\n", retcode);
    goto ex;
  }
  if (warnings)
  {
    HH_GetWarningsText(dev[0], warningstext, warnings);
    printf("\n\n%s", warningstext);
  }

  retcode = HH_SetStopOverflow(dev[0], 0, 10000);	//for example only
  if (retcode < 0)
  {
    printf("\nHH_SetStopOverflow error %d. Aborted.\n", retcode);
    goto ex;
  }


  while (cmd != 'q')
  {
    fflush(stdout);

    HH_ClearHistMem(dev[0]);
    if (retcode < 0)
    {
      printf("\nHH_ClearHistMem error %d. Aborted.\n", retcode);
      goto ex;
    }

    printf("\npress RETURN to start measurement");
    getchar();

    retcode = HH_GetSyncRate(dev[0], &Syncrate);
    if (retcode < 0)
    {
      printf("\nHH_GetSyncRate error %d. Aborted.\n", retcode);
      goto ex;
    }
    printf("\nSyncrate=%1d/s", Syncrate);

    for (i = 0; i < NumChannels; i++)	// for all channels
    {
      retcode = HH_GetCountRate(dev[0], i, &Countrate);
      if (retcode < 0)
      {
        printf("\nHH_GetCountRate error %d. Aborted.\n", retcode);
        goto ex;
      }
      printf("\nCountrate[%1d]=%1d/s", i, Countrate);
    }

   	//here you could check for warnings again

    retcode = HH_StartMeas(dev[0], Tacq);
    if (retcode < 0)
    {
      printf("\nHH_StartMeas error %d. Aborted.\n", retcode);
      goto ex;
    }

    printf("\n\nMeasuring for %1d milliseconds...", Tacq);

    ctcstatus = 0;
    while (ctcstatus == 0)
    {
      retcode = HH_CTCStatus(dev[0], &ctcstatus);
      if (retcode < 0)
      {
        printf("\nHH_CTCStatus error %d. Aborted.\n", retcode);
        goto ex;
      }
    }

    retcode = HH_StopMeas(dev[0]);
    if (retcode < 0)
    {
      printf("\nHH_StopMeas error %1d. Aborted.\n", retcode);
      goto ex;
    }

    printf("\n");
    for (i = 0; i < NumChannels; i++)	// for all channels
    {
      retcode = HH_GetHistogram(dev[0], counts[i], i, 0);
      if (retcode < 0)
      {
        printf("\nHH_GetHistogram error %1d. Aborted.\n", retcode);
        goto ex;
      }

      Integralcount = 0;
      for (j = 0; j < HistLen; j++)
        Integralcount += counts[i][j];

      printf("\n  Integralcount[%1d]=%1.0lf", i, Integralcount);
    }
    printf("\n");

    retcode = HH_GetFlags(dev[0], &flags);
    if (retcode < 0)
    {
      printf("\nHH_GetFlags error %1d. Aborted.\n", flags);
      goto ex;
    }

    if (flags & FLAG_OVERFLOW) printf("\n  Overflow.");

    printf("\nEnter c to continue or q to quit and save the count data.");
    cmd = getchar();
    getchar();
  }

  for (j = 0; j < HistLen; j++)
  {
    for (i = 0; i < NumChannels; i++)
      fprintf(fpout, "%5d ", counts[i][j]);
    fprintf(fpout, "\n");
  }

ex:

  for (i = 0; i < MAXDEVNUM; i++)	//no harm to close all
  {
    HH_CloseDevice(i);
  }
  if (fpout) fclose(fpout);
  printf("\npress RETURN to exit");
  getchar();

  return 0;
}
